internal/controllers/clustercatalog_controller.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"io/fs"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	catalogd "github.com/operator-framework/catalogd/api/core/v1alpha1"
+)
+
+type CatalogCacheFetcherRemover interface {
+	FetchCatalogContents(ctx context.Context, catalog *catalogd.ClusterCatalog) (fs.FS, error)
+	Remove(catalogName string) error
+}
+
+// ClusterCatalogReconciler reconciles a ClusterCatalog object
+type ClusterCatalogReconciler struct {
+	client.Client
+	Cache CatalogCacheFetcherRemover
+}
+
+//+kubebuilder:rbac:groups=olm.operatorframework.io,resources=clustercatalogs,verbs=get;list;watch
+
+func (r *ClusterCatalogReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	existingCatalog := &catalogd.ClusterCatalog{}
+	err := r.Client.Get(ctx, req.NamespacedName, existingCatalog)
+	if client.IgnoreNotFound(err) != nil {
+		return ctrl.Result{}, err
+	}
+	if apierrors.IsNotFound(err) {
+		return ctrl.Result{}, r.Cache.Remove(req.Name)
+	}
+
+	_, err = r.Cache.FetchCatalogContents(ctx, existingCatalog)
+	return ctrl.Result{}, err
```

Comment:
One benefit of the previous approach is that catalog pull errors will show up in ClusterExtension status.

If we move catalog cache populate to a separate controller, do users lose all visibility into the specific details of the reason the catalog caches aren't populated?

I wonder if we should store these errors somewhere that is available to the ClusterExtension reconciler so that we can still propagate them into ClusterExtension status?

Reply:
I'm considering keeping on-demand cache population. If cache fetch fails here - we will keep re-queueing until it gets populated. If it doesn't and it is something permanent - cache population will fail during `ClusterExtension` itself and it will be propagated to `ClusterExtension` status as it currently works.

It should work, but I don't like it very much because it is not straightforward to understand.

> I wonder if we should store these errors somewhere that is available to the ClusterExtension reconciler so that we can still propagate them into ClusterExtension status?

Any ideas? We can probably store somewhere in `filesystemCache` ([e.g. in `cacheDataByCatalogName`](https://github.com/operator-framework/operator-controller/blob/f3e2ce0727136bbfe572e750fef1ce1e38795552/internal/catalogmetadata/cache/cache.go#L54))?

Reply:
High level idea would be for the interface that `ClusterExtension` uses to remain the same (e.g. it can return an error).

Yeah `cacheDataByCatalogName` seems like the obvious place. We could have `cacheData` include an `Err` field?

Reply:
@joelanford could you please take a look at https://github.com/operator-framework/operator-controller/pull/1318?

I refactored the client and cache: the client is now responsible for http requests and cache is responsible for caching only (previously it was making request). And I also added error propagation.

Once it is merged I'll be able to rebase this PR on top of the changes.